bin/varnishd/cache/cache_wrk.c

```diff
@@ -398,10 +398,32 @@ Pool_Work_Thread(struct pool *pp, struct worker *wrk)
 				i = Lck_CondWait(&wrk->cond, &pp->mtx, tmo);
 				if (i == ETIMEDOUT)
 					VCL_Rel(&wrk->vcl);
-			} while (wrk->task->func == NULL);
-			tpx = *wrk->task;
-			tp = &tpx;
-			wrk->stats->summs++;
+				if (wrk->task->func != NULL) {
+					/* We have been handed a new task */
+					tpx = *wrk->task;
+					tp = &tpx;
+					wrk->stats->summs++;
+				} else if (pp->b_stat != NULL &&
```

Comment:
I see the point and I am not against it, but it would be nice if we could maybe refactor a bit more to collapse the two `pool_stat_summ` cases. Wouldn't something similar to
```
if (i == ETIMEDOUT) {
    VCL_Rel(&wrk->vcl);
    Lck_Unlock(&pp->mtx);
    continue;
}
```
at L399 work as well?

Reply:
From bugwash:

> \< martin> that whole loop is getting a bit unwieldy

I agree but haven't tried anything yet to improve that aspect of the change,